pkg/deploy/lattice/service_network_manager.go

```diff
@@ -274,22 +285,78 @@ func (m *defaultServiceNetworkManager) isServiceNetworkAssociatedWithVPC(ctx con
 					glog.V(6).Infoln("ServiceNetwork and Vpc association is being deleted, please retry later")
-					return true, r.Id, resp, errors.New(LATTICE_RETRY)
+					return true, r, resp, errors.New(LATTICE_RETRY)
 				case vpclattice.ServiceNetworkVpcAssociationStatusCreateInProgress:
 					glog.V(6).Infoln("ServiceNetwork and Vpc association is being created, please retry later")
-					return true, r.Id, resp, errors.New(LATTICE_RETRY)
+					return true, r, resp, errors.New(LATTICE_RETRY)
 				}
 			}
 		}
 	}
 	return false, nil, resp, err
 }
+
+func (m *defaultServiceNetworkManager) UpdateServiceNetworkVpcAssociationIfNeeded(ctx context.Context, snId string, snArn string, desiredSN *latticemodel.ServiceNetwork, existingSnvaId *string) (latticemodel.ServiceNetworkStatus, error) {
+	retrievedSnva, err := m.cloud.Lattice().GetServiceNetworkVpcAssociationWithContext(ctx, &vpclattice.GetServiceNetworkVpcAssociationInput{
+		ServiceNetworkVpcAssociationIdentifier: existingSnvaId,
+	})
+	if err != nil {
+		return latticemodel.ServiceNetworkStatus{}, err
+	}
+
+	if len(desiredSN.Spec.SecurityGroupIds) == 0 && retrievedSnva.SecurityGroupIds != nil && len(retrievedSnva.SecurityGroupIds) > 0 {
+		// VPC latiice API have a limitation that the user can't remove all security groups if current snva already has security groups.
+		// If that case happen, the controller will not delete snva. User should manually delete it.
+		//https://docs.aws.amazon.com/vpc-lattice/latest/ug/security-groups.html#security-groups-rules
+
+		// TODO: Add validation webhook or update gateway (or VpcAssociationPolicy) status to indicate that error.
+		//  https://gateway-api.sigs.k8s.io/geps/gep-713/#standard-status-condition
+		return latticemodel.ServiceNetworkStatus{
+			ServiceNetworkID:     snId,
+			ServiceNetworkARN:    snArn,
+			SnvaSecurityGroupIds: nil,
+		}, errors.New(LATTICE_RETRY)
+	}
+
+	twoSlicesElementsEqual := utils.PointerStringSlicesElementsEqual(desiredSN.Spec.SecurityGroupIds, retrievedSnva.SecurityGroupIds)
+
+	if twoSlicesElementsEqual {
+		// desiredSN's security group ids are same with retrievedSnva's security group ids, don't need to update
+		return latticemodel.ServiceNetworkStatus{
+			ServiceNetworkID:     snId,
+			ServiceNetworkARN:    snArn,
+			SnvaSecurityGroupIds: retrievedSnva.SecurityGroupIds,
+		}, nil
+	}
+
+	updateSnvaResp, err := m.cloud.Lattice().UpdateServiceNetworkVpcAssociationWithContext(ctx, &vpclattice.UpdateServiceNetworkVpcAssociationInput{
+		ServiceNetworkVpcAssociationIdentifier: existingSnvaId,
+		SecurityGroupIds:                       desiredSN.Spec.SecurityGroupIds,
+	})
+	retSnStatus := latticemodel.ServiceNetworkStatus{
+		ServiceNetworkID:     snId,
+		ServiceNetworkARN:    snArn,
+		SnvaSecurityGroupIds: updateSnvaResp.SecurityGroupIds,
+	}
+	if err != nil {
+		return retSnStatus, err
+	}
+	switch updateSnvaResp.Status {
+	case aws.String(vpclattice.ServiceNetworkVpcAssociationStatusActive):
```

Comment:
I would reverse condition. If status is not active return error, and return normal response at the end of funciton